patterns/interpreter.md

```diff
@@ -0,0 +1,317 @@
+# Interpreter
+
+## Description
+
+If a problem occurs very often and requires long and repetitive steps to solve it,
+then the problem instances might be expressed in a simple
+language and an interpreter object could solve it by
+interpreting the sentences written in this simple language.
+Basically, for any kind of problems we define a domain
+language, then define a grammar for this language and
+design interpreter solving problem instances.
+
+## Motivation
+
+Imagine that our work is translating simple mathematical expressions into
+[assembly language](https://en.wikipedia.org/wiki/Assembly_language)
+(more simple and low level programming language).
+For simplicity, our expressions consists of ten digits `0`,...,`9`,
+four operations `+, -, /, *` and a pair of parenthesis `(, )`.
+For example, expression `2 + 4` could be translated into
+
+```ignore
+mov eax, 2
+mov ebx, 4
+add eax, ebx
+```
+
+Our goal is to automate translating into assembly instructions
+using the Interpreter design pattern. In other words, we want simply
+provide the Interpreter with an expression and get Assembly
+language output. For example
+
+```rust, ignore
+x.interpret("7+3*(2-1)", &output);
+```
+
+## Context Free Grammars
+
```

Comment:
```suggestion

In order to understand the Interpreter pattern you need to know a little bit
about formal language theory.
If you already know what a grammar is feel free to skip this section,
otherwise fasten your seat belts.

```
Do you like this?